### PR TITLE
Add TCP server timeout

### PR DESCRIPTION
**Pull request by SergiySW**
Make timeouts configurable

### CODE REVIEW

**Comment by cryptocode** [comment]
I don't think start() needs a param after this change, that function should just set cutoff to `now()` (maybe rename from cutoff to start_time or something?)

**Comment by SergiySW** [comment]
@cryptocode good point!

File: nano/node/nodeconfig.cpp [hunk 1]
```diff
@@ -246,6 +250,12 @@ bool nano::node_config::upgrade_json (unsigned version_a, nano::jsonconfig & jso
 			upgraded = true;
 		}
 		case 16:
+		{
+			json.put ("tcp_client_timeout", tcp_client_timeout.count ());
+			json.put ("tcp_server_timeout", tcp_server_timeout.count ());
+			upgraded = true;
```
Comment (wezrule) [inline nano/node/nodeconfig.cpp]: Should this have a test? I added one in a PR but it's not reviewed yet, feel free to copy and replace with your options: https://github.com/nanocurrency/nano-node/pull/1734/files

File: nano/node/bootstrap.cpp [hunk 2]
```diff
@@ -2253,6 +2258,41 @@ void nano::bootstrap_server::finish_request ()
 	{
 		run_next ();
 	}
+	else
+	{
+		auto this_l (shared_from_this ());
+		std::weak_ptr<nano::bootstrap_server> this_w (shared_from_this ());
+		node->alarm.add (std::chrono::steady_clock::now () + node->config.tcp_server_timeout + std::chrono::seconds (1), [this_w]() {
```
Comment (wezrule) [inline nano/node/bootstrap.cpp]: do we need this +1 second?

File: nano/node/bootstrap.cpp [hunk 3]
```diff
@@ -2253,6 +2258,41 @@ void nano::bootstrap_server::finish_request ()
 	{
 		run_next ();
 	}
+	else
+	{
+		auto this_l (shared_from_this ());
```
Comment (wezrule) [inline nano/node/bootstrap.cpp]: is this variable needed?

File: nano/core_test/network.cpp [hunk 4]
```diff
@@ -1423,3 +1423,69 @@ TEST (bootstrap, keepalive)
...
+		ASSERT_EQ (node0->bootstrap.connections.size (), 1);
+	}
+	bool disconnected (false);
+	while (!disconnected)
+	{
+		{
+			std::lock_guard<std::mutex> guard (node0->bootstrap.mutex);
+			disconnected = node0->bootstrap.connections.empty ();
+		}
+		ASSERT_NO_ERROR (system.poll ());
```
[the remark this answers is not in the document]
Comment (SergiySW) [reply nano/core_test/network.cpp]: Probably

File: nano/core_test/network.cpp [hunk 5]
```diff
@@ -1423,3 +1423,69 @@ TEST (bootstrap, keepalive)
...
+TEST (bootstrap, tcp_listener_timeout_empty)
+{
+	nano::system system (24000, 1);
+	auto node0 (system.nodes[0]);
+	node0->config.tcp_server_timeout = std::chrono::seconds (1);
+	auto socket (std::make_shared<nano::socket> (node0));
+	nano::keepalive keepalive;
+	auto input (keepalive.to_bytes ());
+	std::atomic<bool> connected (false);
+	socket->async_connect (node0->bootstrap.endpoint (), [&input, socket, &connected](boost::system::error_code const & ec) {
```
[the remark this answers is not in the document]
Comment (SergiySW) [reply nano/core_test/network.cpp]: Not needed, right

File: nano/node/nodeconfig.cpp (hunk 1 above)
Comment (SergiySW) [reply nano/node/nodeconfig.cpp]: Thanks

File: nano/node/bootstrap.cpp (hunk 3 above)
Comment (SergiySW) [reply nano/node/bootstrap.cpp]: No need

File: nano/node/bootstrap.cpp (hunk 2 above)
Comment (SergiySW) [reply nano/node/bootstrap.cpp]: I.e. now = 100, tcp_server_timeout = 1. Then we should check after 1 second if 100 + 1 <= now ()
But what if there was some action at timestamp 100.999? Then it still will be timed out at 101 even efter 1 ms left

**Pull request merged**